libs/ebpfnetsh/programs.cpp

```diff
@@ -115,20 +141,35 @@ handle_ebpf_add_program(
         return ERROR_SUPPRESS_OUTPUT;
     }
 
-    // TODO(#188): add option to pin all programs in the object.
     struct bpf_program* program = bpf_program__next(nullptr, object);
     struct bpf_link* link;
-    result = ebpf_program_attach(program, (tags[1].bPresent ? &attach_type : nullptr), nullptr, 0, &link);
+    result = ebpf_program_attach(program, (tags[TYPE_INDEX].bPresent ? &attach_type : nullptr), nullptr, 0, &link);
     if (result != EBPF_SUCCESS) {
         std::cerr << "error " << result << ": could not attach program" << std::endl;
         return ERROR_SUPPRESS_OUTPUT;
     }
 
-    if (!pinned.empty()) {
-        if (bpf_program__pin(program, pinned.c_str()) < 0) {
-            std::cerr << "error " << errno << ": could not pin program" << std::endl;
+    if (pinned_type == PT_FIRST) {
+        if (pinpath.empty()) {
+            pinpath = bpf_program__name(program);
+        }
+        if (bpf_program__pin(program, pinpath.c_str()) < 0) {
+            std::cerr << "error " << errno << ": could not pin to " << pinpath << std::endl;
             return ERROR_SUPPRESS_OUTPUT;
         }
+    } else if (pinned_type == PT_ALL) {
+        // The pinpath specified is like a "directory" under which to pin programs.
+        if (!pinpath.empty()) {
+            pinpath += "/";
+        }
+        bpf_object__for_each_program(program, object)
+        {
+            std::string fullpath = pinpath + bpf_program__name(program);
+            if (bpf_program__pin(program, fullpath.c_str()) < 0) {
+                std::cerr << "error " << errno << ": could not pin to " << fullpath << std::endl;
```

Comment:
If pin of one program fails, I think we should undo everything done before this (program load, attach and pinning of some of the program)

Reply:
the bpftool behavior is to close the object in this case (leaving any successful pins in place).  I will do that and add a comment.

Reply:
Isn't that a leak though? The netsh call fails, but some programs are left loaded and attached and pinned, and the caller will have to make additional netsh calls to cleanup that state. If the ELF file contains multiple programs, then they will have to cleanup / unpin each program one by one